Draft - Add CAIP-2 for NEAR Protocol (#43)
ChainAgnostic/CAIPs: CAIPs/caip-draft_near_protocol.md

---
caip: 28
title: Blockchain Reference for the NEAR Protocol Namespace
author: Aaron Luhning (@ALuhning), Trevor Clarke (@TrevorJTClarke)
discussions-to: https://github.com/ceramicnetwork/js-ceramic/issues/880, https://github.com/ceramicnetwork/js-ceramic/pull/881
status: Draft
type: Standard
created: 2021-02-04
updated: 2021-02-04

> requires 2

Reviewer: ```suggestion
requires: 2
```

---

## Simple Summary

> Provides details of the NEAR Protocol namespace and reference for CAIP-2.

Reviewer: ```suggestion
Defines the NEAR Protocol namespace and reference for CAIP-2.
```

## Abstract

In CAIP-2 a general blockchain identification scheme is defined.  This is the implementation of CAIP-2 for NEAR Protocol.

# Motivation

See CAIP-2.

## Specification

### NEAR Protocol Namespace

The namespace "near" refers to the wider [NEAR Protocol](https://near.org) ecosystem.

#### Reference Definition

> The definition for this namespace will use the `chain_id` as an identifier for each of the three NEAR Protocol chains. The format is a string equal to either `mainnet`, `testnet` or `betanet`.  The chain identifier is appended by a [network id](https://chainid.network/).

Reviewer: ```suggestion
The definition for this namespace will use the `chain_id` as an identifier for each of the three NEAR Protocol chains. The format is a string equal to either `mainnet`, `testnet` or `betanet`.  The chain identifier is appended by a [chain id](https://chainid.network).
```

The chain ID defined in EIP155 is the most widely used chain identifier in the Ethereum ecosystem known to the authors. It strives for uniqueness and the fact that the standard is used for replay protection ensure that creators of a new Ethereum network have an incentive to use an ID that is not used elsewhere.

Since there is EVM compatible shards, EIP155 is specified in addition to the root shard definitions. As more shards enable different capabilities, additional CAIPs will need specification.

## Resolution Method

To obtain the chain_id for the NEAR Protocol namespace, make a JSON-RPC request to the appropriate rpc endpoint with method `EXPERIMENTAL_genesis_config`.
RPC endpoints are: `https://rpc.mainnet.near.org`, `https://rpc.testnet.near.org`, or `https://rpc.betanet.near.org`.  For example (testnet):

```jsonc
// Request
{
  "jsonrpc": "2.0",
  "id": "dontcare",
  "method": "EXPERIMENTAL_genesis_config"
}

// Response
{
  "jsonrpc": "2.0",
  "result": {
    "protocol_version": 29,
    "genesis_time": "2020-07-31T03:39:42.911378Z",

>     "chain_id": "testnet",

Reviewer: not a big fan of this tbh - IMHO "testnet" is a bad ID - especially as in this context a chain_id is usually an EIP155 number

Reviewer 2: Good point -- This is part of the root shard response. We will bring up with the NEAR Core team and see if we can define a better standards payload item/id.

Reviewer 3: Any updates here? Would be good to have this merged :)

Reviewer 4: I think it would be a good idea to merge this as is right now.

The CAIP spec includes a Resolution Method for this namespace which is exactly what is required.

The feedback is related to core logic on Near clients which will reflect any changes through the Resolution Method described so this shouldn't need to block the PR.

Reviewer 5: @ligi Can this be merged? I agree this shouldn't be held up by core logic in NEAR clients.

    "genesis_height": 10885359,
    "num_block_producer_seats": 100,

>     ...

Reviewer: I would add one `}` to have both blocks closed
